### PR TITLE
[Android] autofill service address

### PR DESCRIPTION
**Pull request by tapanmodh**
Resolves https://github.com/brave/brave-browser/issues/36226

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/security/issues/new/choose) them https://github.com/brave/reviews/issues/1520
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Brave should be there in autofill service of settings of the mobile device
- If there is a edit text for address in any other app then user can fill that with address saved in brave app
- User can save the address into brave app directly from other app after user successfully fill the address.

| Screenshot_20240218_145320 | Screenshot_20240218_145808 |
| ------------- | ------------- |

### CODE REVIEW

File: browser/ui/android/strings/android_brave_strings.grd [hunk 1]
```diff
@@ -4102,6 +4102,75 @@ If you don't accept this request, VPN will not reconnect and your internet conne
...
+      container
+    </message>
+    <message name="IDS_USERNAME" desc="Autofill field for username">
+      username
+    </message>
+    <message name="IDS_LOGIN" desc="Autofill field for login">
+      login
+    </message>
+    <message name="IDS_ID" desc="Autofill field for id">
+      id
```
Comment (SergeyZhukovsky) [inline browser/ui/android/strings/android_brave_strings.grd:4144]: do we need to translate it?

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+            @NonNull FillRequest request, @NonNull FillCallback callback) {
+        mFillRequest = request;
+        mFillCallback = callback;
+    }
+
+    public BraveAutofillBackgroundServiceImpl(
+            @NonNull SaveRequest request, @NonNull SaveCallback callback) {
+        mSaveRequest = request;
+        mSaveCallback = callback;
+        mIsSaveRequest = true;
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:63]: Looks like it's always true, do we need it for future use?

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    @Override
+    protected void launchBrowser(Context context, String tag) {
+
+        final BrowserParts parts =
+                new EmptyBrowserParts() {
+                    @Override
+                    public void finishNativeInitialization() {
+                        Context context = ContextUtils.getApplicationContext();
+                        if (context == null) {
+                            if (mIsSaveRequest) mSaveCallback.onSuccess();
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:75]: Lets use brackets for uniformity.

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+        }
+    }
+
+    @Nullable
+    protected String getHint(Context context, @NonNull ViewNode node) {
+        // First try the explicit autofill hints...
+
+        String[] hints = node.getAutofillHints();
+        if (hints != null) {
+            // We're simple, we only care about the first hint
```
Comment (samartnik) [inline android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:397]: Why only first hint btw?

File: browser/ui/android/strings/android_brave_strings.grd [hunk 5]
```diff
@@ -4102,6 +4102,75 @@ If you don't accept this request, VPN will not reconnect and your internet conne
       If your current service provider is insecure, Brave uses a secure alternative DNS provider
     </message>
     <!-- End of Secure DNS Settings -->
+
+    <!-- Autofill Service strings -->
+    <message name="IDS_CITY" desc="Autofill field for city">
```
Comment (samartnik) [inline browser/ui/android/strings/android_brave_strings.grd:4107]: Not sure about these translations. Are hints localised? We can't really guarantee that we will localise them precisely the same way.

File: browser/ui/android/strings/android_brave_strings.grd (hunk 1 above)
Comment (deeppandya) [reply browser/ui/android/strings/android_brave_strings.grd:4144]: yeah we need the translation for different regions as apps would be using the translated IDs.  we need to make sure of it during verification.

File: android/java/AndroidManifest.xml [hunk 6]
```diff
@@ -231,6 +231,16 @@
     </intent-filter>
 </service>
 
+<service
+    android:name="org.chromium.chrome.browser.autofill.BraveAutofillService"
+    android:label="@string/app_name"
+    android:permission="android.permission.BIND_AUTOFILL_SERVICE"
+    android:exported="true">
```
Comment (deeppandya) [inline android/java/AndroidManifest.xml:238]: CC @stoletheminerals as we export the service here.

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillService.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import android.service.autofill.SaveRequest;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.google.android.gms.gcm.TaskParams;
+
+import org.chromium.chrome.browser.init.MinimalBrowserStartupUtils;
+
+@RequiresApi(Build.VERSION_CODES.O)
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/autofill/BraveAutofillService.java:23]: it seems Android 8 (Oreo) is already the minimal required for Brave. Do we really need this annotation?

**Review by AlexeyBarabash — APPROVED** [review]
++

File: android/java/AndroidManifest.xml (hunk 6 above)
Comment (stoletheminerals) [reply android/java/AndroidManifest.xml:238]: Thanks @deeppandya. This is fine, we have to export and `android.permission.BIND_AUTOFILL_SERVICE` is a system-level permission, so only system apps can bind to this service.

File: browser/ui/android/strings/android_brave_strings.grd (hunk 1 above)
Comment (tapanmodh) [reply browser/ui/android/strings/android_brave_strings.grd:4144]: yeah correct

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java (hunk 2 above)
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:63]: there are two constructor with different type of parameters. One for fill request and in this case, it is for save request. so `mIsSaveRequest` variable is used for differentiate between request type in other part of code. 
Ex. https://github.com/brave/brave-core/blob/bd6c097b27b4ed073009c583671dbee5a9fc03a7/android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java#L79C29-L79C43

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java (hunk 3 above)
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:75]: sure, I'll update it

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillService.java (hunk 7 above)
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/autofill/BraveAutofillService.java:23]: yeah correct. It seems Android 8 is already minimum sdk required for Brave. I'll remove the annotation.

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java (hunk 4 above)
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:397]: we are getting autofill hints from node from first hint as of now. I'll work on this with follow up tasks for autofill passwords.

File: android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java (hunk 2 above)
Comment (samartnik) [reply android/java/org/chromium/chrome/browser/autofill/BraveAutofillBackgroundServiceImpl.java:63]: Oh, right, I missed the first c-tor

File: browser/ui/android/strings/android_brave_strings.grd (hunk 1 above)
Comment (mkarolin) [reply browser/ui/android/strings/android_brave_strings.grd:4144]: @deeppandya I am sure the localizers will ask for context, what sort of `id` we are talking about here? Things like passport/driver license/government id card number?

Comment (tapanmodh) [reply browser/ui/android/strings/android_brave_strings.grd:4144]: @mkarolin it is for user id

File: browser/ui/android/strings/android_brave_strings.grd (hunk 5 above)
Comment (tapanmodh) [reply browser/ui/android/strings/android_brave_strings.grd:4107]: these strings are for identification. we are not 100% sure but we have to identify it from hint or label. Maybe we can find something else as well to figure out in follow up work for passwords.

Comment (samartnik) [reply browser/ui/android/strings/android_brave_strings.grd:4107]: yeah, it's a question how reliable our localisations will be for proper fill

**Review by samartnik — APPROVED** [review]
lgtm

**Review by stoletheminerals — APPROVED** [review]

**Review by deeppandya — APPROVED** [review]
Recent changes for company and names LGTM

**Pull request merged**